z3/src/rec_func_decl.rs

```diff
@@ -0,0 +1,184 @@
+use ast;
+use ast::Ast;
+use std::convert::TryInto;
+use std::ffi::CStr;
+use std::fmt;
+use z3_sys::*;
+use {Context, RecFuncDecl, Sort, Symbol, Z3_MUTEX};
+
+impl<'ctx> RecFuncDecl<'ctx> {
+    pub fn new<S: Into<Symbol>>(
+        ctx: &'ctx Context,
+        name: S,
+        domain: &[&Sort<'ctx>],
+        range: &Sort<'ctx>,
+    ) -> Self {
+        assert!(domain.iter().all(|s| s.ctx.z3_ctx == ctx.z3_ctx));
+        assert_eq!(ctx.z3_ctx, range.ctx.z3_ctx);
+
+        let domain: Vec<_> = domain.iter().map(|s| s.z3_sort).collect();
+
+        unsafe {
+            Self::from_raw(
+                ctx,
+                Z3_mk_rec_func_decl(
+                    ctx.z3_ctx,
+                    name.into().as_z3_symbol(ctx),
+                    domain.len().try_into().unwrap(),
+                    domain.as_ptr(),
+                    range.z3_sort,
+                ),
+            )
+        }
+    }
+
+    pub unsafe fn from_raw(ctx: &'ctx Context, z3_func_decl: Z3_func_decl) -> Self {
+        let guard = Z3_MUTEX.lock().unwrap();
+
+        Z3_inc_ref(ctx.z3_ctx, Z3_func_decl_to_ast(ctx.z3_ctx, z3_func_decl));
+
+        Self { ctx, z3_func_decl }
+    }
+
+    /// Adds the body to a recursive function.
+    ///
+    /// ```
+    /// # use z3::{Config, Context, RecFuncDecl, Solver, Sort, Symbol, ast::Int, SatResult};
+    /// # use std::convert::TryInto;
+    /// # let cfg = Config::new();
+    /// # let ctx = Context::new(&cfg);
+    /// let mut f = RecFuncDecl::new(
+    ///     &ctx,
+    ///     "f",
+    ///     &[&Sort::int(&ctx)],
+    ///     &Sort::int(&ctx));
+    /// let n = Int::new_const(&ctx, "n");
+    /// f.add_def(
+    ///     &[&n.clone().into()],
+    ///     &Int::add(&ctx, &[&n, &Int::from_i64(&ctx, 1)])
+    /// );
+    ///
+    /// let f_of_n = &f.apply(&[&n.clone().into()]);
+    ///
+    /// let solver = Solver::new(&ctx);
+    /// let forall: z3::ast::Bool = z3::ast::forall_const(
+    ///         &ctx,
+    ///         &[&n.clone().into()],
+    ///         &[],
+    ///         &n.lt(&f_of_n.as_int().unwrap())
+    ///     ).try_into().unwrap();
+    ///
+    /// solver.assert(&forall);
+    /// let res = solver.check();
+    /// assert_eq!(res, SatResult::Sat);
+    /// ```
+    ///
+    /// Note that `args` should have the types corresponding to the `domain` of the `RecFuncDecl`.
+    pub fn add_def(
+        &self,
+        args: &[&ast::Dynamic<'ctx>],
```

Comment:
do we need to assert on args & body being in the same ctx?

Reply:
I can't comment on this, I don't really know where/how one would use multiple contexts.